pytest_ethereum/utils/linker.py

```diff
@@ -0,0 +1,96 @@
+from eth_utils import to_canonical_address, to_hex
+from eth_utils.toolz import assoc_in
+
+from ethpm.utils.chains import (
+    check_if_chain_matches_chain_uri,
+    create_block_uri,
+    get_chain_id,
+)
+from pytest_ethereum.exceptions import LinkerError
+
+
+def pluck_matching_uri(deployment_data, w3):
+    """
+    Return any blockchain uri that matches w3-connected chain, if one
+    is present in the deployment data keys.
+    """
+    for uri in deployment_data.keys():
+        if check_if_chain_matches_chain_uri(w3, uri):
+            return uri
+    raise LinkerError(
+        "No matching blockchain URI found in deployment_data: {0}, for w3 instance: {1}.".format(
+            list(deployment_data.keys()), w3.__repr__()
+        )
+    )
+
+
+def contains_matching_uri(deployment_data, w3):
+    """
+    Returns true if any blockchain uri in deployment data matches
+    w3-connected chain.
+    """
+    for uri in deployment_data.keys():
+        if check_if_chain_matches_chain_uri(w3, uri):
+            return True
+    return False
+
+
+def create_latest_block_uri(w3, tx_receipt):
+    """
+    Creates a new block uri from data in w3 and provided tx_receipt.
+    """
+    chain_id = get_chain_id(w3)
+    block_hash = to_hex(tx_receipt.blockHash)
+    return create_block_uri(to_hex(chain_id), block_hash)
+
+
+def insert_new_deployment(package, deployment_insert, latest_block_uri):
+    """
+    Returns a new manifest. If a matching chain uri is found in the old manifest, it will
+    update the chain uri along with the new deployment data. If no match, it will simply add
+    the new chain uri and deployment data.
+    """
+    if "deployments" not in package.package_data:
+        return assoc_in(
+            package.package_data, ("deployments", latest_block_uri), deployment_insert
+        )
+
+    old_deployment_data = package.package_data["deployments"]
+    if contains_matching_uri(old_deployment_data, package.w3):
+        old_chain_uri = pluck_matching_uri(old_deployment_data, package.w3)
+        old_chain_data = old_deployment_data.pop(old_chain_uri)
```

Comment:
This is mutating the `package.package_data['deployments']` value, you should probably just do a regular dictionary lookup.